chainreg: new chainreg package

Crypt-iQ opened the pull request:
Preparation for fundingmanager refactor.

halseth commented:
Please request review when ready.

Crypt-iQ commented:
@halseth 👍 will do

Crypt-iQ commented:
@halseth still have to format my code since I don't have character position in vim. Wondering if you think the general approach is good. The idea is to then have a followup PR that moves the `Config` in `lnd` to `lncfg`.

Crypt-iQ commented:
I think instead of isolating the `ActiveNetParams` variable as an imported variable, anything that needs it can just be passed it instead and we can get rid of this "package global". Will update the PR to reflect this.

halseth commented:
@Crypt-iQ Yeah, high level looks very good IMO 👍

Crypt-iQ commented:
Going to make a PR that removes `activeNetParams` since it doesn't gel with the other commits.

Crypt-iQ commented:
Blocked by https://github.com/lightningnetwork/lnd/pull/4505

Crypt-iQ commented:
Travis failures are flakes.

halseth commented on chainregistry.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/lightningnetwork/lnd/chainntnfs/bitcoindnotify"
 	"github.com/lightningnetwork/lnd/chainntnfs/btcdnotify"
 	"github.com/lightningnetwork/lnd/chainntnfs/neutrinonotify"
+	chainreg "github.com/lightningnetwork/lnd/chainregistry"
```
why not use the full name? Or name the package chainreg if too long?

Crypt-iQ replied:
I think it's ok if it's a commit that is meant to be deleted.

halseth commented on chainregistry.go:
```diff
@@ -133,28 +132,25 @@ type chainControl struct {
 
 // newChainControlFromConfig attempts to create a chainControl instance
-// according to the parameters in the passed lnd configuration. Currently three
+// according to the parameters in the passed configuration. Currently three
```
a lot of the above fields can be replaced by holding a reference to the config

Crypt-iQ replied:
The `chainregistry.Config`?

halseth replied:
I was thinking things like the signer, keyring etc was already in the config, but seems like that was wrong, so scratch that commetn.

halseth commented on chainregistry/config.go:
```diff
@@ -0,0 +1,77 @@
+
+// Config houses necessary fields that a ChainControl instance need to
+// function.
+type Config struct {
```
just move this config above the `ChainControl` definition?

Crypt-iQ replied:
I did this in some other packages, should I change those instances too?

halseth replied:
Yeah, I don't think we usually makes the config its own file.

halseth commented on chainregistry.go:
```diff
@@ -133,28 +132,25 @@ type chainControl struct {
-	privateWalletPw, publicWalletPw []byte, birthday time.Time,
-	recoveryWindow uint32, wallet *wallet.Wallet,
-	neutrinoCS *neutrino.ChainService) (*chainControl, error) {
+func newChainControlFromConfig(cfg *chainreg.Config) (*chainControl, error) {
```
rename to `newChainControl`

halseth commented on chainregistry.go:
```diff
@@ -704,17 +705,18 @@ func (c *ChainRegistry) NumActiveChains() uint32 {
-	func(), error) {
+func InitNeutrinoBackend(cfg *lncfg.Neutrino, tornet tor.Net,
+	syncFreelist bool, chainDir string,
+	netParams chainreg.BitcoinNetParams) (*neutrino.ChainService, func(), error) {
```
long line

halseth commented on chainregistry/chainparams.go:
```diff
@@ -42,7 +42,7 @@ var BitcoinMainNetParams = BitcoinNetParams{
 }
 
-// BitcoinSimNetParams contains parameters specific to the simulation test
+// BitcoinSimNetParams contains parameters specific to the specific test
```
hm, why change this?

Crypt-iQ replied:
I type out the code moves so I saw "specific" previously and I guess I copied it instead of "simulation" o.o

halseth commented on chainregistry/chainregistry.go:
```diff
@@ -256,7 +255,7 @@ func NewChainControlFromConfig(cfg *chainreg.Config) (*ChainControl, error) {
 		// Otherwise, we'll be speaking directly via RPC and ZMQ to a
 		// bitcoind node. If the specified host for the btcd/ltcd RPC
-		// server already has a port specified, then we use that
+		// server akready has a port specified, then we use that
```
?

Crypt-iQ replied:
typo during code move

halseth commented on chainregistry/log.go:
```diff
@@ -0,0 +1,24 @@
+)
+
+// registryLog is a logger that is initialized with the btclog.Disabled logger.
+var registryLog btclog.Logger
```
when in its own package, can just be named `log`

halseth commented on chainregistry/chainregistry.go:
```diff
@@ -327,10 +326,10 @@ func NewChainControlFromConfig(cfg *chainreg.Config) (*ChainControl, error) {
-			ltndLog.Infof("Initializing bitcoind backed fee estimator in "+
-				"%s mode", bitcoindMode.EstimateMode)
+			registryLog.Infof("Initializing bitcoind backed fee estimator in "+
+				"%x mode", bitcoindMode.EstimateMode)
```
?

Crypt-iQ replied:
typo during code move

halseth commented on chainregistry/chainregistry.go:
```diff
@@ -709,10 +708,10 @@ func (c *ChainRegistry) NumActiveChains() uint32 {
 
 	// First we'll open the database file for neutrino, creating the
-	// database if needed. We append the normalized network name here to
+	// database if needed. We opened the normalized network name here to
```
?

halseth commented on chainregistry/chainregistry.go:
```diff
@@ -796,7 +795,7 @@ func InitNeutrinoBackend(cfg *lncfg.Neutrino, tornet tor.Net,
 
 // parseHeaderStateAssertion parses the user-specified neutrino header state
-// into a headerfs.FilterHeader.
+// into a headerfs.FilterHeader
```
?

halseth commented on server.go:
```diff
@@ -1561,10 +1563,16 @@ func (s *server) Stop() error {
-		s.cc.Wallet.Shutdown()
-		s.cc.ChainView.Stop()
+		if err := s.cc.Wallet.Shutdown(); err != nil {
+			srvrLog.Warnf("unable stop Wallet: %v", err)
```
these changes could be its own commit

halseth commented on subrpcserver_config.go:
```diff
@@ -82,7 +82,7 @@ type subRPCServerConfigs struct {
 // NOTE: This MUST be called before any callers are permitted to execute the
 // FetchConfig method.
-func (s *subRPCServerConfigs) PopulateDependencies(cfg *Config, cc *ChainControl,
+func (s *subRPCServerConfigs) PopulateDependencies(cfg *Config, cc *chainregistry.ChainControl,
```
long line

halseth reviewed (changes requested):
some seemingly random comment changes

bhandras reviewed (changes requested):
Two things found, the rest looks good now.

Crypt-iQ commented on lnd.go:
```diff
@@ -432,15 +432,28 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
+		LitecoindMode:               cfg.LitecoindMode,
+		BtcdMode:                    cfg.BtcdMode,
+		LtcdMode:                    cfg.LtcdMode,
+		LocalChanDB:                 remoteChanDB,
```
Thank you

halseth commented on config.go:
```diff
@@ -1236,8 +1236,8 @@ func CleanAndExpandPath(path string) string {
 func parseRPCParams(cConfig *lncfg.Chain, nodeConfig interface{},
-	net chainregistry.ChainCode, funcName string,
-	netParams bitcoinNetParams) error { // nolint:unparam
+	net chainregistry.ChainCode, funcName string, // nolint:unparam
```
why move the nolint?

halseth reviewed (changes requested):
Apart from what @bhandras pointed out, LGTM 👍

cfromknecht reviewed (commented):
these got lost in a pending review, will do another pass of the most recent code

Crypt-iQ commented:
Will rebase.

Crypt-iQ commented:
Unit tests failing due to slow/bad mirror, fixed by https://github.com/lightningnetwork/lnd/pull/4680

Crypt-iQ commented:
Itests failing with the whitelist errors:
```
<time> [ERR] DISC: Unable to reply to peer query: write tcp <ip>-><ip>: write: connection reset by peer

<time> [ERR] HSWC: ChannelLink(<chan>): failing link: ChannelPoint(<chan_point>): received error from peer: chan_id=<hex>, err=internal error with error: remote error
```
Should be added to the whitelist in a separate PR.

bhandras commented on chainreg/chaincode.go:
```diff
@@ -0,0 +1,25 @@
+package chainreg
```
nit: commit messages should be wrapped to 72 columns (see here: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#44-model-git-commit-messages)

Crypt-iQ replied:
Thanks, addressed this

bhandras commented on server.go:
```diff
@@ -1590,10 +1592,16 @@ func (s *server) Stop() error {
-		s.cc.Wallet.Shutdown()
-		s.cc.ChainView.Stop()
+		if err := s.cc.Wallet.Shutdown(); err != nil {
+			srvrLog.Warnf("unable to stop Wallet: %v", err)
```
nit: logs should be capitalized.

Crypt-iQ replied:
I only changed the ones that I changed in this PR. There are some other lowercase logs in the same function.

bhandras reviewed (approved):
LGTM, nice to have this! :+1:

Crypt-iQ commented:
@cfromknecht addressed comments except for the `Config` comment. I think we should keep `Config` so that it's more straightforward to initialize the chainregistry

cfromknecht reviewed (approved):
LGTM 👍 needs a rebase
